Combinator/Combinator.fs

```diff
@@ -167,14 +167,26 @@ module Combinator =
     let manyTill p pEnd = manyTillWithCount 0 p pEnd
 
     let manyTill1 p pEnd = manyTillWithCount 1 p pEnd
-        
-    let manyN num (parser) =                              
-        takeTillWithCount 0 (fun i -> List.length i > num) parser    >>= fun result -> 
-            if result.Length <> num then
-                failwith ("Error, attempted to match " + num.ToString() + " but only got " + result.Length.ToString())
 
-            preturn result          
-    
+
+
+    let manyN num parser state =
+        let didFind found currentState =
+            match List.length found with
+            | x when x > 0 -> Some (List.rev found), currentState
+            | _ -> None, currentState
+        let rec many' parser (resultList, currentState: State<_,_,_>, cnt) =
+            let returnValue acc = didFind acc currentState
+            match currentState.hasMore() with
+            | false -> returnValue resultList
+            | true ->
+                match parser currentState with
+                | Some (result), nextState ->
+                    if cnt >= num then returnValue (result :: resultList)
```

Comment:
Should this be equal to and not just greater than or equal to?  We want to not accidentally parse too many items, it specifically needs to be N items or fail